platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/bls/protocol/CounterpartyManager.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.bls.protocol;
+
+import com.swirlds.common.system.NodeId;
+import java.util.*;
+
+/**
+ * Object which keeps track of the behavior and status of counterparties during BLS protocol
+ * execution
+ */
+public class CounterpartyManager {
+    /** Set of counterparties who behaved in a verifiably incorrect way, assumed to be malicious */
+    private final Set<NodeId> maliciousNodes;
+
+    /** Set of counterparties who failed to send an expected message */
+    private final Set<NodeId> offlineNodes;
+
+    /** List of reports detailing the disqualification of a counterparty */
+    private final List<IncidentReport> incidentReports;
+
+    /** Constructor */
+    public CounterpartyManager() {
+        this.maliciousNodes = new HashSet<>();
+        this.offlineNodes = new HashSet<>();
+
+        this.incidentReports = new ArrayList<>();
+    }
+
+    /**
+     * Add a malicious party
+     *
+     * @param nodeId the id of the malicious party
+     */
+    public void declareMalicious(final NodeId nodeId, final IncidentReport report) {
+        // Don't disqualify a party in multiple ways
+        if (checkDisqualified(nodeId)) {
+            return;
+        }
+
+        maliciousNodes.add(nodeId);
+        incidentReports.add(report);
+    }
+
+    /**
+     * Add an offline party
+     *
+     * @param nodeId the id of the offline party
+     */
+    public void declareOffline(final NodeId nodeId, final IncidentReport report) {
```

Comment:
Is there a way for `offline` parties to come back online?   Or is this an irreversible state?   What is the process for restoring contact with offline parties and bringing them back into good standing?

Reply:
By definition, a party declared to be `offline` has arrived at an irrecoverable state, by failing to send a required message.

If a node is declared offline by other nodes, it will have to wait until the next rekeying to participate.